extensions/s3/src/main/java/io/deephaven/extensions/s3/S3RequestCache.java

```diff
@@ -59,27 +57,36 @@ int getFragmentSize() {
      * @return the request
      */
     @NotNull
-    Request.RequestInfo getOrCreateRequest(@NotNull final S3Uri uri, final long fragmentIndex,
+    S3ChannelContext.Request.AcquiredRequest getOrCreateRequest(@NotNull final S3Uri uri, final long fragmentIndex,
             @NotNull final S3ChannelContext context) {
-        final Mutable<Request.RequestInfo> ret = new MutableObject<>();
-        // TODO Need to unwrap the compute to avoid putting() the same result in the map.
-        requests.compute(new Request.ID(uri, fragmentIndex), (key, existingRequest) -> {
+        final Request.ID key = new Request.ID(uri, fragmentIndex);
+        Request.AcquiredRequest newAcquiredRequest = null;
+        Request existingRequest = requests.get(key);
+        for (int retryCount = 0; retryCount < Integer.MAX_VALUE; retryCount++) {
             if (existingRequest != null) {
-                final Request.RequestInfo acquiredExisting = existingRequest.tryAcquire();
-                if (acquiredExisting != null) {
-                    ret.setValue(acquiredExisting);
-                    return existingRequest;
+                final Request.AcquiredRequest acquired = existingRequest.tryAcquire();
+                if (acquired != null) {
+                    return acquired;
+                } else {
+                    remove(existingRequest);
                 }
             }
-            final Request.RequestInfo newRequestInfo = Request.createAndAcquire(fragmentIndex, context);
-            ret.setValue(newRequestInfo);
-            if (log.isDebugEnabled()) {
-                log.debug().append("Adding new request to cache: ").append(String.format("ctx=%d ",
-                        System.identityHashCode(context))).append(newRequestInfo.request.requestStr()).endl();
+            if (newAcquiredRequest == null) {
+                newAcquiredRequest = Request.createAndAcquire(fragmentIndex, context);
             }
-            return newRequestInfo.request;
-        });
-        return ret.getValue();
+            existingRequest = requests.putIfAbsent(key, newAcquiredRequest.request);
```

Comment:
I think it would be better to use `replace` (followed by `get` on failure) rather than `remove` + `putIfAbsent`. However, I'm suspicious that `KeyedObjectHash.replace` has a bug and needs to be fixed to enable that slight optimization.

Reply:
The `replace` code does have a bug in case of a `null`. I have added an branch for the two cases, and a comment explaining the reasoning. Let me know if this is better.